Temperature on App Icon via Badge

lprokein opened the pull request:
https://github.com/erikmartens/NearbyWeather/issues/30

lprokein commented:
Please see https://github.com/erikmartens/NearbyWeather/issues/30#issuecomment-431689333 before reviewing.

erikmartens commented:
Hey @lprokein, thanks for your work! I added the label for you, have to see how I can tweak the permissions. Please allow me to take a few days for completing the review. I'll probably only get to doing it on the weekend.

lprokein commented:
Hey, thanks for the review. I updated my code based on your suggestions. I removed alternate icons, some unnecessary remote notifications code, unified permissions handling, reworked preferred bookmark settings based on your settings dialog and some other updates based on your review.

I used this "hacky" picker because it can handle many options by scrolling but now I've tested that your solution can also handle greater amount of options so I used that instead to make sure it doesn't break in future.

Btw why use `unowned` instead of `weak` here https://github.com/erikmartens/NearbyWeather/pull/47#discussion_r228763216 ?

erikmartens commented:
@lprokein Sorry, I've been very busy, hence I'm only getting back to you now. I hope I'll be able to re-review the PR tonight or tomorrow evening and answer your question then. And after that we shall finally merge this PR :+1:

erikmartens commented on NearbyWeather/PreferencesManager.swift:
```diff
@@ -16,6 +16,74 @@ public protocol PreferencesOption {
     var stringValue: String { get }
 }
 
+public struct LightCityStruct: Codable {
```
rename to `PreferredBookmark`

erikmartens commented on NearbyWeather/PreferencesManager.swift:
```diff
@@ -16,6 +16,74 @@ public protocol PreferencesOption {
+    }
+}
+
+public class PreferredBookmark: Codable, PreferencesOption {
```
I think you are making this a bit more complicated than it needs to be. Why don't you just make `var value` (and then also rename it accordingly) optional and just store the id of the bookmarked city in it? Then you don't need the `LightCityStruct`, nor the `PreferredBookmarkWrappedEnum`. For the `var stringValue: String` you can just look up the city name via the data base and if it is nil you can return `R.string.localizable.none()`.

erikmartens commented on NearbyWeather/SettingsTableViewController.swift:
```diff
@@ -21,6 +25,11 @@ class SettingsTableViewController: UITableViewController {
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
         
+        BadgeService.shared.areBadgesEnabled { [weak self] enabled in
```
I don't think the cell should be hidden. If we do the notification permissions the way we should, then there's a point where they are not determined yet. Then this would also be hidden and the user could never activate the feature. I think what needs to be done when it is detected that there no notification permissions is to just set the toggle to off (and also falsify the preference behind). The preferred city should be untouched and the user should always be able to set it, regardless of whether they are actually receiving the badge.

lprokein commented:
Hey, I fixed some issues you addressed in your review. I hope it's fine now. I am afraid I want be able to do more fixes this time because my school exams are coming so I won't have more time.

I also resolved merge conflicts and updated my code to Swift 4.2 as I seen that you been also updating code to 4.2.

My native language is Slovak, I can do translation later on via separate PR.

erikmartens commented:
Hey @lprokein, sorry to keep you waiting. Uni is also getting a bit more intense right now for me. I have exams in January. I'll have much more time in February. There's no rush with the translation or fixing the PR. I'll review it now and will give you feedback.

erikmartens commented:
Hey @lprokein, how is it going? Would you like to continue working on this? If I don't hear back I'll finish this up myself at some point, but I'll add you as a contributor of course. Thank you for your work so far.

lprokein commented:
Hi @erikmartens , sorry I can't promise when I will have time to finish this I got too much busy now with school, work and personal projects.

erikmartens commented on NearbyWeather/WeatherDataManager.swift:
```diff
@@ -160,7 +169,56 @@ class WeatherDataManager {
+                return
+            }
+            
+            if let bookmarkIdx = self.bookmarkedWeatherDataObjects?.firstIndex(where: { $0.locationId == unwrappedPreferredBookmarkWeatherData.locationId }) {
```
Done

erikmartens commented on NearbyWeather/WeatherDataManager.swift:
```diff
@@ -160,7 +169,56 @@ class WeatherDataManager {
                 NotificationCenter.default.post(name: Notification.Name(rawValue: kWeatherServiceDidUpdate), object: self)
-                completionHandler?()
+                BadgeService.shared.updateBadge() {
+                    completionHandler?(.success)
```
Done

erikmartens commented on NearbyWeather/BadgeService.swift:
```diff
@@ -0,0 +1,131 @@
+        BadgeService.shared.updateBadge(withCompletionHandler: nil)
+    }
+    
+    public func updateBadge(withCompletionHandler completionHandler: (() -> ())?) {
```
Done

erikmartens commented on NearbyWeather/WeatherDataManager.swift:
```diff
@@ -160,7 +169,56 @@ class WeatherDataManager {
+            DispatchQueue.main.async {
+                NotificationCenter.default.post(name: Notification.Name(rawValue: kWeatherServiceDidUpdate), object: self)
+                BadgeService.shared.updateBadge() {
+                    completionHandler(.success)
```
Done

erikmartens commented on NearbyWeather/AppDelegate.swift:
```diff
@@ -80,6 +84,26 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         refreshWeatherDataIfNeeded()
     }
     
+    func application(_ application: UIApplication, didRegister notificationSettings: UIUserNotificationSettings) {
```
Done

erikmartens commented on NearbyWeather/PermissionsManager.swift:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+    
+    public func didRegisterNotificationSettings(_ settings: UIUserNotificationSettings) {
```
Done

erikmartens commented on NearbyWeather/BadgeService.swift:
```diff
@@ -0,0 +1,131 @@
+        }
+    }
+    
+    public func setBadgeServiceEnabled(_ enabled: Bool) {
```
Done

erikmartens commented on NearbyWeather/PermissionsManager.swift:
```diff
@@ -0,0 +1,71 @@
+        callBadgePermissionsCompletion(withSuccess: settings.types.contains(UIUserNotificationType.badge))
+    }
+    
+    private func callBadgePermissionsCompletion(withSuccess success: Bool) {
```
Done

erikmartens commented on NearbyWeather/PermissionsManager.swift:
```diff
@@ -0,0 +1,71 @@
+    // MARK: - Properties
+    
+    public static var shared: PermissionsManager!
+    private var notificationPermissionsRequestCompletion: ((Bool) -> ())?
```
Done
